ec/src/pairing.rs

```diff
@@ -0,0 +1,201 @@
+    }
+}
+
+/// Represents the target group of a pairing. This struct is a 
+/// wrapper around the field that the target group is embedded in.
+#[derive(Derivative, CanonicalSerialize, CanonicalDeserialize)]
+#[derivative(Copy(bound = ""), Clone(bound = ""), PartialEq(bound = ""), Eq(bound = ""), Debug(bound = ""), Default(bound = ""), Hash(bound = ""))]
+pub struct PairingOutput<P: Pairing>(P::TargetField);
+
+impl<P: Pairing> PairingOutput<P> {
+    /// Converts `self` into an element of the underlying field.
+    pub fn to_field_element(&self) -> P::TargetField {
+        self.0
+    }
+}
+
+impl<P: Pairing> Display for PairingOutput<P> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl<P: Pairing> Zero for PairingOutput<P> {
+    /// The identity element, or "zero", of the group is the identity element of the multiplicative group of the underlying field, i.e., `P::TargetField::one()`.
+    fn zero() -> Self {
+        Self(P::TargetField::one())
+    }
+
+    fn is_zero(&self) -> bool {
+        self.0.is_one()
+    }
+}
+
+impl<'a, P: Pairing> Add<&'a Self> for PairingOutput<P> {
+    type Output = Self;
+
+    #[inline]
+    fn add(mut self, other: &'a Self) -> Self {
+        self += other;
+        self
+    }
+}
+
+impl<'a, P: Pairing> AddAssign<&'a Self> for PairingOutput<P> {
+    fn add_assign(&mut self, other: &'a Self) {
+        self.0 *= other.0;
+    }
+}
+
+impl<'a, P: Pairing> SubAssign<&'a Self> for PairingOutput<P> {
+    fn sub_assign(&mut self, other: &'a Self) {
+        self.0 *= other.0.inverse().unwrap();
+    }
+}
+
+impl<'a, P: Pairing> Sub<&'a Self> for PairingOutput<P> {
+    type Output = Self;
+
+    #[inline]
+    fn sub(mut self, other: &'a Self) -> Self {
+        self -= other;
+        self
+    }
+}
+
+ark_ff::impl_additive_ops_from_ref!(PairingOutput, Pairing);
+
+impl<P: Pairing> MulAssign<P::ScalarField> for PairingOutput<P> {
+    fn mul_assign(&mut self, other: P::ScalarField) {
+        *self = *self * other.into_repr()
+    }
+}
+
+impl<P: Pairing> Mul<<P::ScalarField as PrimeField>::BigInt> for PairingOutput<P> {
+    type Output = Self;
+    fn mul(mut self, other: <P::ScalarField as PrimeField>::BigInt) -> Self {
+        self.0 = self.0.pow(other.as_ref());
+        self
+    }
+}
+
+impl<P: Pairing> Zeroize for PairingOutput<P> {
+    fn zeroize(&mut self) {
+        self.0.zeroize()
+    }
+}
+
+impl<P: Pairing> Neg for PairingOutput<P> {
+    type Output = Self;
+
+    #[inline]
+    fn neg(self) -> Self {
+        Self(self.0.inverse().unwrap())
+    }
+}
+
+impl<P: Pairing> Distribution<PairingOutput<P>> for Standard {
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> PairingOutput<P> {
+        // Sample a random G1 element
+        let g1 = P::G1::rand(rng);
+        // Sample a random G2 element
+        let g2 = P::G2::rand(rng);
+        P::pairing(g1, g2)
```

Comment:
Sampling a random GT element requires sampling random g1 and g2 elements, and then pairing them.

Reply:
I wonder if this impacts anything..

Reply:
Does it not suffice to sample a random GT element, and then take the final exponentiation on it?

Reply:
That's a good idea, it should work. That can also provide us with a generator, right?

Reply:
It'd only provide a generators independent of pairing outputs, but most protocols need a generator e(G_1,G_2) if they need a G_t generator at all.